src/Client/ClientBase.h

```diff
@@ -235,6 +236,36 @@ class ClientBase : public Poco::Util::Application, public IHints<2, ClientBase>
     } profile_events;
 
     QueryProcessingStage::Enum query_processing_stage;
+
+    struct HostPort
+    {
+        String host;
+        std::optional<UInt16> port{};
+        friend std::istream & operator>>(std::istream & in, HostPort & hostPort)
+        {
+            String host_with_port;
+            in >> host_with_port;
+            DB::DNSResolver & resolver = DB::DNSResolver::instance();
+            try
+            {
+                Poco::Net::SocketAddress address = resolver.resolveAddress(host_with_port);
+                hostPort.host = address.host().toString();
+                hostPort.port = address.port();
+            }
+            catch (const Exception & e)
+            {
+                if (e.message() == "Missing port number")
+                {
+                    hostPort.host = resolver.resolveHost(host_with_port).toString();
+                    hostPort.port = std::nullopt;
+                    return in;
+                }
+                throw;
+            }
```

Comment:
It doesn't work for IPv6 host without port:
```
./clickhouse client --host 2001:3984:3989::1:1000
Code: 36. DB::Exception: Port must be numeric. (BAD_ARGUMENTS)
```
And in general checking port existance by throwing and catching an exception is not good. We can manually check if the hostname is an IPv6 address (for example by checking that we have more then one colon in the host name) and if it is, we can check if the host name contains the port number by checking the host name format (it should be like [<IPv6 address>]:port). If it's not an IPv6 address, we also can check if the host name contains the port number. We can even create a new method for `DNSResolver` for this purpose (smth like `resolveHostOrAddress`).

Reply:
Yes, I understood this solution doesn't work for IPv6 only after committing. I thought, that I need to change static function `splitHostAndPort` in `DNSResolver.cpp`, because it throws wrong exception, when it gets IPv6 host (it throws `Port must be numeric` when port wasn't set at all). But your solution with `resolveHostOrAddress` looks better than catching exception, I will implement `resolveHostOrAddress` method.